### PR TITLE
Repacked the intellij-agent artifact to kover-jvm-agent

### LINKED ISSUE #464: Repack the intellij-agent artifact to kover-jvm-agent
In the case where the `kover-cli` is used for generating reports, users may need to perform on the fly instrumentation that will produce compatible results.

Since `intellij-agent` has a versioning that differs from the version of Kover, it will be problematic for the user to choose the appropriate version.
It can also confuse the use of dependencies from intellij, which is not explicitly related to Kover for the user.

Additionally, in the new artifact, we may pass parameters and system variables in a different form and with different names, more typical for Kover

### PR DESCRIPTION
**Pull request by shanshin**
User can use the JVM agent separately, with the name kover and without mentioning IntelliJ, as well as reduce the number of parameters and properties required to transfer to the agent.

Now Kover Gradle Plugin uses kover-jvm-agent with the same version. To do this, the transfer of staging repositories was added to the functional tests. The common logic of publishing was also extracted into a separate convention plugin.

Resolves #464

### CODE REVIEW

File: build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-fat-jar-conventions.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+val fatJarDependency = "fatJar"
+val fatJarConfiguration = configurations.create(fatJarDependency)
+
+
+tasks.jar {
+    from(
+        fatJarConfiguration.map { if (it.isDirectory) it else zipTree(it) }
+    ) {
+        exclude("OSGI-OPT/**")
```
Comment (sandwwraith) [inline build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-fat-jar-conventions.gradle.kts:33]: Do we really have something OSGI-related? If not, let's do not add unused excludes in order to avoid confusion

File: kover-cli/build.gradle.kts [hunk 2]
```diff
@@ -23,7 +23,6 @@ plugins {
 
 extensions.configure<Kover_publishing_conventions_gradle.KoverPublicationExtension> {
     description.set("Command Line Interface for Kotlin Coverage Toolchain")
-    fatJar.set(true)
 }
```
Comment (sandwwraith) [inline kover-cli/build.gradle.kts:26]: Why CLI shouldn't be a fat jar anymore? Will users still be able to do `java -jar kover-cli.jar ...`?

File: kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/plugin/dsl/KoverVersions.kt [hunk 3]
```diff
@@ -21,4 +23,10 @@ public object KoverVersions {
      * JaCoCo coverage tool version used by default.
      */
     public const val JACOCO_TOOL_DEFAULT_VERSION = "0.8.10"
+
+    /**
+     * Current version of Kover Gradle Plugin
+     */
+    public val version: String
+        get() = KoverFeatures.getVersion()
```
Comment (sandwwraith) [inline kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/plugin/dsl/KoverVersions.kt:31]: Are versions for Features artifact and Gradle plugin will always be the same?

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    public static void premain(String argsString, Instrumentation instrumentation) throws Exception {
+        KoverAgentSettings settings = readSettingsFromFile(extractArgsFile(argsString));
+        IntellijIntegration.callPremain(settings, instrumentation);
+    }
+
+    private static File extractArgsFile(String args) {
+        if (!args.startsWith(FILE_PREFIX_IN_ARGS)) {
+            throw new IllegalArgumentException("Incorrect arguments format for Kover JVM agent, arguments are expected to start with " + FILE_PREFIX_IN_ARGS);
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java:37]: Are you sure that allowing passing arguments only in file and not accepting them in CLI is a good idea? Maybe some users don't want them to be in file?

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/IntellijIntegration.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        CoveragePremain.premain(intelliJArgsString, instrumentation);
+    }
+
+    private static void setIntellijSystemProperties() {
+        System.setProperty("idea.coverage.use.system.classloader", "true");
+
+        // Disable agent logging to stdout for messages of levels `debug`, `info`, `warn`.
+        System.setProperty("idea.coverage.log.level", "error");
+
+        // Enables saving the array in the candy field,
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/IntellijIntegration.java:44]: you mean cOndy field? don't remember any candies in class files :(

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/IntellijIntegration.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * A flag to calculate coverage for unloaded classes.
+     */
+    private static final boolean CALCULATE_FOR_UNLOADED_CLASSES = false;
+
+    /**
+     * Create hit block only for line, false adds the ability to count branches
+     */
+    private static final boolean LINING_ONLY_MODE = false;
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/IntellijIntegration.java:25]: Tbh `LINES_ONLY_MODE` sounds better than `LINING_ONLY_MODE`

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverAgentSettings.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package kotlinx.kover.jvmagent;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class KoverAgentSettings {
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverAgentSettings.java:6]: It seems that this class does not need equals/hashCode, but maybe add `.toString()` for better debuggability?

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/ParseUtils.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package kotlinx.kover.jvmagent;
+
+public class ParseUtils {
+    private ParseUtils() {
+        // no-op
+    }
+
+    public static boolean isBoolean(String value) {
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/ParseUtils.java:8]: I think we can safely move this to `KoverJvmAgentPremain` without creating unnecessary utils class

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                } else if (line.startsWith(APPEND_ATTRIBUTE)) {
+                    String value = line.substring(APPEND_ATTRIBUTE.length());
+                    if (!isBoolean(value)) {
+                        throw new IllegalArgumentException("Incorrect value for argument " + APPEND_ATTRIBUTE + " in Kover JVM agent arguments file, expected true or false");
+                    }
+                    settings.appendToReportFile = Boolean.parseBoolean(value);
+                } else if (line.length() == 0) {
+                    // skip empty line
+                } else {
+                    throw new IllegalArgumentException("Unrecognized line in Kover arguments file: " + line);
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java:73]: If args file is supposed to be filled by user (and from commit message I suppose it is), it would be nice to mention in the exception what valid values are. E.g. `Line must start with one of prefixes: ${listOf(FILE_PATH_ATTRIBUTE, APPEND_ATTRIBUTE, ...}`

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                try {
+                    reader.close();
+                } catch (IOException ignored) {
+                    // no-op
+                }
+            }
+        }
+
+        if (settings.reportFilePath == null) {
+            throw new IllegalArgumentException("Path to the report file is required, add " + FILE_PATH_ATTRIBUTE + " attribute to the args file");
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java:89]: Shouldn't we also check that `File(reportFilePath).exists()` or is it done later in the agent?

File: build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-publishing-conventions.gradle.kts [hunk 11]
```diff
@@ -24,19 +21,100 @@ plugins {
...
+tasks.register<CollectTask>("collectRepository") {
+    dependsOn(publicationTask)
+    dependsOn(externalSnapshots)
+
+    local.convention(extension.localRepository)
+    externals.from(externalSnapshots)
+}
+
+@CacheableTask
+abstract class CollectTask: DefaultTask() {
```
Comment (sandwwraith) [inline build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-publishing-conventions.gradle.kts:86]: I do not see instances and configuration of this task, am I missing something?

File: settings.gradle.kts [hunk 12]
```diff
@@ -8,12 +6,19 @@ pluginManagement {
     }
 }
 
+plugins {
+    id("org.gradle.toolchains.foojay-resolver-convention") version("0.7.0")
```
Comment (sandwwraith) [inline settings.gradle.kts:10]: Is this plugin related to the change? I do not see configured toolchains

**Review by sandwwraith — CHANGES_REQUESTED** [review]
If the Kover agent is supposed to be used by end users, some documentation for it is needed (at least for argsfile setup)

File: kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/plugin/dsl/KoverVersions.kt (hunk 3 above)
Comment (shanshin) [reply kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/plugin/dsl/KoverVersions.kt:31]: Yes

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java (hunk 4 above)
Comment (shanshin) [reply kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java:37]: In the future, we can add it, but so far there has not been such a request, and I do not want to come up with a format ahead of time

File: build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-fat-jar-conventions.gradle.kts (hunk 1 above)
Comment (shanshin) [reply build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-fat-jar-conventions.gradle.kts:33]: Similar directories were in IntelliJ dependencies

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java (hunk 10 above)
Comment (shanshin) [reply kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/KoverJvmAgentPremain.java:89]: Agent will create file if it isn't exist

File: build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-publishing-conventions.gradle.kts (hunk 11 above)
Comment (shanshin) [reply build-logic/src/main/kotlin/kotlinx/kover/conventions/kover-publishing-conventions.gradle.kts:86]: `tasks.register<CollectTask>("collectRepository") {` above

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/IntellijIntegration.java (hunk 5 above)
Comment (shanshin) [reply kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/IntellijIntegration.java:44]: 🍬

File: kover-cli/build.gradle.kts (hunk 2 above)
Comment (sandwwraith) [reply kover-cli/build.gradle.kts:27]: I just don't see it in this buildscript. Is it added by default by conventions?

Comment (shanshin) [reply kover-cli/build.gradle.kts:27]: Yes, applying fat jar convention plugin allows you to create archives if you use the `fatJarConfiguration`

File: kover-jvm-agent/docs/index.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+## On the fly instrumentation
+On the fly instrumentation - modification of the bytecode of classes in order to measure coverage, that occurs when the class is loaded into the JVM.
+
+To instrument the loaded classes, it is necessary to connect a JVM agent to the Java application being launched.
+
+## Getting a readable report
+To get a readable coverage report, you need to:
+1. [Connect JVM agent on JVM start](#connecting-the-jvm-agent)
+2. Terminate the JVM application. When application finished, the binary report file will be saved
```
Comment (sandwwraith) [inline kover-jvm-agent/docs/index.md:11]:
```suggestion
2. Run your JVM application and wait until it exits. When the application is finished, the binary report file will be saved.
```

(also add dots for other items)

File: kover-jvm-agent/docs/index.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+3. Using the [Kover CLI](/cli#generating-reports), generate a human-readable report from a binary report
```
Comment (sandwwraith) [inline kover-jvm-agent/docs/index.md:12]: Is `cli/#...` link correct? It is not resolvable in IntelliJ

File: kover-jvm-agent/docs/index.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+1. Download the latest version of the Kover JVM agent jar from [maven repository](https://mvnrepository.com/artifact/org.jetbrains.kotlinx/kover-jvm-agent).
+2. Put JVM agent jar file in a local directory (Important! renaming a jar file is not allowed)
+3. Create a file with agent arguments, [learn more about the arguments](#kover-jvm-arguments-file)
+4. Add an argument to the java application startup command `-javaagent:<path_to_agent_jar>=file:<path_to_settings_file>`.
+   Example of an application launch command `java -jar application.jar -javaagent:/opt/kover-jvm-agent-0.7.6.jar=file:/tmp/agent.args`
+
+## Kover JVM arguments file
+The arguments file is a set of settings, each of which is written on a new line.
+Line format: `argument_name=argument_value`
+
```
Comment (sandwwraith) [inline kover-jvm-agent/docs/index.md:24]: Please add info like `report.file argument is required, while the rest are optional`.

File: kover-jvm-agent/docs/index.md [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+List of arguments:
+- `report.file` - path to the file, which will contain a binary coverage report in ic format. The file is created if it did not exist before
+- `report.append` - it is acceptable to specify true or false. if true, then if the file will be appended if the coverage is already stored in it
+- `exclude` - specify which classes do not need to be modified when loading. For such classes, the coverage will always be 0.
+  
+  It is acceptable to use `*` and `?` wildcards, `*` means any number of arbitrary characters (including no chars), `?` means one arbitrary character.
+- `exclude.regex` - specify which classes do not need to be modified when loading. For such classes, the coverage will always be 0.
```
Comment (sandwwraith) [inline kover-jvm-agent/docs/index.md:31]: Is `exclude.regex` simply combined with `exclude` or something else happens? Please add info to the doc, e.g. "It is possible to use `exclude` and `exclude.regex` at the same time"

File: kover-jvm-agent/docs/index.md [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+  It is acceptable to specify regex.
+- `include` - specify which classes will be modified when loading, all other classes will not.
+
+  It is acceptable to use `*` and `?` wildcards, `*` means any number of arbitrary characters (including no chars), `?` means one arbitrary character.
+- `include.regex` - specify which classes will be modified when loading, all other classes will not.
+
+  It is acceptable to specify regex.
+
+Example of arguments file
```
Comment (sandwwraith) [inline kover-jvm-agent/docs/index.md:41]:
```suggestion
Example of arguments file:
```

File: kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/ParseUtils.java [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package kotlinx.kover.jvmagent;
+
+public class ParseUtils {
```
Comment (sandwwraith) [inline kover-jvm-agent/src/main/java/kotlinx/kover/jvmagent/ParseUtils.java:3]: I believe this file can be deleted completely

File: kover-jvm-agent/docs/index.md (hunk 14 above)
Comment (shanshin) [reply kover-jvm-agent/docs/index.md:12]: It should be correct after moving the documentation to a single location

**Review by sandwwraith — APPROVED** [review]
